contracts/utils/HStrategyHelper.sol

```diff
@@ -0,0 +1,508 @@
+    /// @return token1Amount amount of token1 needed to be pulled from uniV3
+    function calculateExtraTokenAmountsForMoneyVault(
+        IIntegrationVault moneyVault,
+        HStrategy.TokenAmounts memory expectedTokenAmounts
+    ) external view returns (uint256 token0Amount, uint256 token1Amount) {
+        (uint256[] memory minTvl, uint256[] memory maxTvl) = moneyVault.tvl();
+        token0Amount = (minTvl[0] + maxTvl[0]) >> 1;
+        token1Amount = (minTvl[1] + maxTvl[1]) >> 1;
+
+        if (token0Amount > expectedTokenAmounts.moneyToken0) {
+            token0Amount -= expectedTokenAmounts.moneyToken0;
+        } else {
+            token0Amount = 0;
+        }
+
+        if (token1Amount > expectedTokenAmounts.moneyToken1) {
+            token1Amount -= expectedTokenAmounts.moneyToken1;
+        } else {
+            token1Amount = 0;
+        }
+    }
+
+    /// @notice calculates expected amounts of tokens after rebalance
+    /// @param expectedRatios ratios of the capital on different assets
+    /// @param expectedTokenAmountsInToken0 expected capitals (in token0) on the strategy vaults
+    /// @param domainPositionParams current position and pool state combined with predictions from the oracle
+    /// @return amounts amounts of tokens expected after rebalance on the strategy vaults
+    function calculateExpectedTokenAmounts(
+        HStrategy.ExpectedRatios memory expectedRatios,
+        HStrategy.TokenAmountsInToken0 memory expectedTokenAmountsInToken0,
+        HStrategy.DomainPositionParams memory domainPositionParams
+    ) external pure returns (HStrategy.TokenAmounts memory amounts) {
+        amounts.erc20Token0 = FullMath.mulDiv(
+            expectedRatios.token0RatioD,
+            expectedTokenAmountsInToken0.erc20TokensAmountInToken0,
+            expectedRatios.token0RatioD + expectedRatios.token1RatioD
+        );
+        amounts.erc20Token1 = FullMath.mulDiv(
+            expectedTokenAmountsInToken0.erc20TokensAmountInToken0 - amounts.erc20Token0,
+            domainPositionParams.spotPriceX96,
+            CommonLibrary.Q96
+        );
+
+        amounts.moneyToken0 = FullMath.mulDiv(
+            expectedRatios.token0RatioD,
+            expectedTokenAmountsInToken0.moneyTokensAmountInToken0,
+            expectedRatios.token0RatioD + expectedRatios.token1RatioD
+        );
+        amounts.moneyToken1 = FullMath.mulDiv(
+            expectedTokenAmountsInToken0.moneyTokensAmountInToken0 - amounts.moneyToken0,
+            domainPositionParams.spotPriceX96,
+            CommonLibrary.Q96
+        );
+        {
+            uint256 uniCapitalRatioX96 = FullMath.mulDiv(
+                FullMath.mulDiv(
+                    domainPositionParams.spotPriceSqrtX96 - domainPositionParams.lowerPriceSqrtX96,
+                    CommonLibrary.Q96,
+                    domainPositionParams.upperPriceSqrtX96 - domainPositionParams.spotPriceSqrtX96
```

Comment:
I don't get how this formula works, but domainPositionParams.upperPriceSqrtX96 = domainPositionParams.spotPriceSqrtX96 is possible as I know? So this is zero division? Also, could you please provide an explanation of the formula?

Reply:
There is the formula to calculate the ratio between tokens in the uniV3 position. Here's the actual bug with zero division, in the border case, because the actual ratio of tokens in the position can be infinite. We'll fix it

Reply:
There is the formula to calculate the ratio between tokens in the uniV3 position. Here's the actual bug with zero division, in the border case, because the actual ratio of tokens in the position can be infinite. We'll fix it